### DISCUSSION TITLE
Install support through Poetry and atlernative build systems

### DISCUSSION THREAD

**Discussion by t03i** [networkit/networkit#775] [Ideas]
Installing networkit is not yet possible through poetry. From my understanding, this is due to the encapsulated build process of poetry. Specifically, Cython support through poetry seems to be best achieved using PEP 518 and specifying build dependencies or a build script in a pyproject.toml file [1] [2]. 

Implementing PEP 518 would not only be a QoL improvement for poetry users, it could also be an initial step into adopting poetry for networkit in general. This would significantly simplify python dependency management, as well as package building and publishing.

**Comment by fabratu** [comment]
Thanks for the request and the initial research on how to get support for poetry.

We have to look into it. In reality supporting a package manager always comes with a cost in maintaining stuff. From what I read, poetry is quite popular to keep these costs low - but there is always a tradeoff in supporting more systems. It seems you can mix conda and poetry if the dependencies are not too complicated.

I will turn this in a discussion btw. - since it is not really issue, more something we could improve on.

**Comment by appleparan** [comment]
In my case, I could install `networkit` via `pyenv`+`poetry`. 

```
$ PYTHON_CONFIGURE_OPTS="--enable-shared" pyenv install 3.9.6
...
$ poetry add cmake, Cython
$ poetry add networkit
$ poetry run python
Python 3.9.6 (default, Aug 16 2021, 09:17:13)
[GCC 9.3.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import networkit
>>>
```

Here is my `pyproject.toml`

```
[tool.poetry]
...

[tool.poetry.dependencies]
python = ">=3.7.1, <3.11"
torch = "^1.9.0"
torchvision = "^0.10.0"
torchaudio = "^0.9.0"
numpy = "^1.21.2"
pandas = "^1.3.2"
cmake = "^3.21.1"
Cython = "^0.29.24"
networkit = "^9.0"

[tool.poetry.dev-dependencies]
pytest = "^5.2"

[build-system]
requires = ["poetry-core>=1.0.0a9"]
build-backend = "poetry.core.masonry.api"
```

**Reply by t03i** [reply]
I can confirm that  the installation seems to work even without `pyenv` using  `Poetry 1.1.8, Python 3.96, clang 12.0.1, g++ 11.1.0, arch linux`.
It seems like the changes to make the compilation work on windows also positively affected the poetry installation. 
This requires a bit more testing (python versions, systems) but seems otherwise implemented.
